File: app/src/main/java/org/tvheadend/tvhclient/ui/features/channels/ChannelListFragment.java
package org.tvheadend.tvhclient.ui.features.channels;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.ProgressBar;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.Channel;
import org.tvheadend.tvhclient.domain.entity.ChannelTag;
import org.tvheadend.tvhclient.domain.entity.Connection;
import org.tvheadend.tvhclient.domain.entity.Program;
import org.tvheadend.tvhclient.domain.entity.Recording;
import org.tvheadend.tvhclient.ui.base.BaseFragment;
import org.tvheadend.tvhclient.ui.base.callbacks.RecyclerViewClickCallback;
import org.tvheadend.tvhclient.ui.features.dialogs.ChannelTagSelectionDialog;
import org.tvheadend.tvhclient.ui.features.dialogs.GenreColorDialog;
import org.tvheadend.tvhclient.ui.features.dvr.RecordingAddEditActivity;
import org.tvheadend.tvhclient.ui.features.programs.ProgramListFragment;
import org.tvheadend.tvhclient.ui.features.search.SearchRequestInterface;
import org.tvheadend.tvhclient.util.menu.PopupMenuUtil;
import org.tvheadend.tvhclient.util.menu.SearchMenuUtils;
import org.tvheadend.tvhclient.util.tasks.WakeOnLanTask;

import java.util.List;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import timber.log.Timber;

public class ChannelListFragment extends BaseFragment implements RecyclerViewClickCallback, ChannelDisplayOptionListener, SearchRequestInterface, Filter.FilterListener {

    @BindView(R.id.recycler_view)
    protected RecyclerView recyclerView;
    @BindView(R.id.progress_bar)
    protected ProgressBar progressBar;

    private ChannelRecyclerViewAdapter recyclerViewAdapter;
    private ChannelViewModel viewModel;
    private int selectedTimeOffset;
    private int selectedListPosition;
    private String searchQuery;
    private Unbinder unbinder;

    // Used in the time selection dialog to show a time entry every x hours.
    private final int intervalInHours = 2;
    private int programIdToBeEditedWhenBeingRecorded = 0;
    private List<ChannelTag> channelTags;
    private long selectedTime;

    private Runnable currentTimeUpdateTask;
    private final Handler currentTimeUpdateHandler = new Handler();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.recyclerview_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onDestroyView() {
        recyclerView.setAdapter(null);
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState != null) {
            selectedListPosition = savedInstanceState.getInt("listPosition", 0);
            selectedTimeOffset = savedInstanceState.getInt("timeOffset");
            searchQuery = savedInstanceState.getString(SearchManager.QUERY);
        } else {
            selectedListPosition = 0;
            selectedTimeOffset = 0;
            Bundle bundle = getArguments();
            if (bundle != null) {
                searchQuery = bundle.getString(SearchManager.QUERY);
            }
        }

        recyclerViewAdapter = new ChannelRecyclerViewAdapter(isDualPane, this);
        recyclerView.setLayoutManager(new LinearLayoutManager(activity.getApplicationContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(activity.getApplicationContext(), LinearLayoutManager.VERTICAL));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(recyclerViewAdapter);

        recyclerView.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);

        viewModel = ViewModelProviders.of(activity).get(ChannelViewModel.class);

        Timber.d("Observing selected time");
        viewModel.getSelectedTime().observe(getViewLifecycleOwner(), time -> {
            Timber.d("View model returned selected time " + time);
            if (time != null) {
                selectedTime = time;
            }
        });

        Timber.d("Observing channel tags");
        viewModel.getChannelTags().observe(getViewLifecycleOwner(), tags -> {
            if (tags != null) {
                Timber.d("View model returned " + tags.size() + " channel tags");
                channelTags = tags;
            }
        });

        Timber.d("Observing channels");
        viewModel.getChannels().observe(getViewLifecycleOwner(), channels -> {
            if (channels != null) {
                Timber.d("View model returned " + channels.size() + " channels");
                recyclerViewAdapter.addItems(channels);
            }
            if (recyclerView != null) {
                recyclerView.setVisibility(View.VISIBLE);
            }
            if (progressBar != null) {
                progressBar.setVisibility(View.GONE);
            }
            showChannelTagOrChannelCount();

            if (isDualPane && recyclerViewAdapter.getItemCount() > 0) {
                showChannelDetails(selectedListPosition);
            }
        });

        // Get all recordings for the given channel to check if it belongs to a certain program
        // so the recording status of the particular program can be updated. This is required
        // because the programs are not updated automatically when recordings change.
        Timber.d("Observing recordings");
        viewModel.getAllRecordings().observe(getViewLifecycleOwner(), recordings -> {
            if (recordings != null) {
                Timber.d("View model returned " + recordings.size() + " recordings");
                recyclerViewAdapter.addRecordings(recordings);
                for (Recording recording : recordings) {
                    // Show the edit recording screen of the scheduled recording
                    // in case the user has selected the record and edit menu item.
                    if (recording.getEventId() == programIdToBeEditedWhenBeingRecorded
                            && programIdToBeEditedWhenBeingRecorded > 0) {
                        programIdToBeEditedWhenBeingRecorded = 0;
                        Intent intent = new Intent(activity, RecordingAddEditActivity.class);
                        intent.putExtra("id", recording.getId());
                        intent.putExtra("type", "recording");
                        activity.startActivity(intent);
                        break;
                    }
                }
            }
        });

        // Initiate a timer that will update the view model data every minute
        // so that the progress bars will be displayed correctly
        currentTimeUpdateTask = () -> {
            long currentTime = System.currentTimeMillis();
            Timber.d("Checking if selected time " + selectedTime + " is past current time " + currentTime);
            if (selectedTime < currentTime) {
                Timber.d("Updated selected time to current time");
                viewModel.setSelectedTime(currentTime);
            }
            currentTimeUpdateHandler.postDelayed(currentTimeUpdateTask, 60000);
        };
        currentTimeUpdateHandler.post(currentTimeUpdateTask);
    }

    private void showChannelTagOrChannelCount() {
        // Show either all channels or the name of the selected
        // channel tag and the channel count in the toolbar
        String toolbarTitle = viewModel.getSelectedChannelTagName(activity);
        if (TextUtils.isEmpty(searchQuery)) {
            toolbarInterface.setTitle(toolbarTitle);
            toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.items,
                    recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
        } else {
            toolbarInterface.setTitle(getString(R.string.search_results));
            toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.channels,
                    recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("listPosition", selectedListPosition);
        outState.putInt("timeOffset", selectedTimeOffset);
        outState.putString(SearchManager.QUERY, searchQuery);
    }

    @Override
    public void onResume() {
        super.onResume();
        // When the user returns from the settings only the onResume method is called, not the
        // onActivityCreated, so we need to check if any values that affect the representation
        // of the channel list have changed.
        //noinspection ConstantConditions
        viewModel.setChannelSortOrder(Integer.valueOf(sharedPreferences.getString("channel_sort_order", activity.getResources().getString(R.string.pref_default_channel_sort_order))));
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.channel_list_options_menu, menu);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        final boolean showGenreColors = sharedPreferences.getBoolean("genre_colors_for_channels_enabled",
                activity.getResources().getBoolean(R.bool.pref_default_genre_colors_for_channels_enabled));
        final boolean showChannelTagMenu = sharedPreferences.getBoolean("channel_tag_menu_enabled",
                activity.getResources().getBoolean(R.bool.pref_default_channel_tag_menu_enabled));

        if (TextUtils.isEmpty(searchQuery)) {
            menu.findItem(R.id.menu_genre_color_info_channels).setVisible(showGenreColors);
            menu.findItem(R.id.menu_timeframe).setVisible(isUnlocked);
            menu.findItem(R.id.menu_search).setVisible((recyclerViewAdapter.getItemCount() > 0));

            // Prevent the channel tag menu item from going into the overlay menu
            if (showChannelTagMenu) {
                menu.findItem(R.id.menu_tags).setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
            }
        } else {
            menu.findItem(R.id.menu_genre_color_info_channels).setVisible(false);
            menu.findItem(R.id.menu_timeframe).setVisible(false);
            menu.findItem(R.id.menu_search).setVisible(false);
            menu.findItem(R.id.menu_tags).setVisible(false);
        }

        menu.findItem(R.id.menu_wol).setVisible(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_tags:
                return ChannelTagSelectionDialog.showDialog(activity, channelTags, appRepository.getChannelData().getItems().size(), this);

            case R.id.menu_timeframe:
                return menuUtils.handleMenuTimeSelection(selectedTimeOffset, intervalInHours, 12, this);

            case R.id.menu_genre_color_info_channels:
                return GenreColorDialog.showDialog(activity);

            case R.id.menu_sort_order:
                return menuUtils.handleMenuChannelSortOrderSelection(this);

            case R.id.menu_wol:
                Connection connection = appRepository.getConnectionData().getActiveItem();
                new WakeOnLanTask(activity, connection).execute();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onTimeSelected(int which) {
        selectedTimeOffset = which;
        if (recyclerView != null) {
            recyclerView.setVisibility(View.GONE);
        }
        if (progressBar != null) {
            progressBar.setVisibility(View.VISIBLE);
        }
        // Add the selected list index as extra hours to the current time.
        // If the first index was selected then use the current time.
        long timeInMillis = System.currentTimeMillis();
        timeInMillis += (1000 * 60 * 60 * which * intervalInHours);
        viewModel.setSelectedTime(timeInMillis);
    }

    @Override
    public void onChannelTagIdsSelected(@NonNull Set<Integer> ids) {
        if (recyclerView != null) {
            recyclerView.setVisibility(View.GONE);
        }
        if (progressBar != null) {
            progressBar.setVisibility(View.VISIBLE);
        }
        viewModel.setSelectedChannelTagIds(ids);
    }

    @Override
    public void onChannelSortOrderSelected(int id) {
        viewModel.setChannelSortOrder(id);
    }

    /**
     * Show the program list when a channel was selected. In single pane mode a separate
     * activity is called. In dual pane mode a list fragment will be shown in the right side.
     *
     * @param position The selected position in the list
     */
    private void showChannelDetails(int position) {
        selectedListPosition = position;
        recyclerViewAdapter.setPosition(position);
        Channel channel = recyclerViewAdapter.getItem(position);
        if (channel == null || !isVisible()
                || !activity.getLifecycle().getCurrentState().isAtLeast(Lifecycle.State.STARTED)) {
            return;
        }

        FragmentManager fm = activity.getSupportFragmentManager();
        if (!isDualPane) {
            // Show the fragment to display the program list of the selected channel.
            Bundle bundle = new Bundle();
            bundle.putString("channelName", channel.getName());
            bundle.putInt("channelId", channel.getId());
            bundle.putLong("selectedTime", selectedTime);

            ProgramListFragment fragment = new ProgramListFragment();
            fragment.setArguments(bundle);
            fm.beginTransaction()
                    .replace(R.id.main, fragment)
                    .addToBackStack(null)
                    .commit();
        } else {
            // Check if an instance of the program list fragment for the selected channel is
            // already available. If an instance exist already then update the selected time
            // that was selected from the channel list.
            Fragment fragment = fm.findFragmentById(R.id.details);
            if (!(fragment instanceof ProgramListFragment)
                    || ((ProgramListFragment) fragment).getShownChannelId() != channel.getId()) {
                fragment = ProgramListFragment.newInstance(channel.getName(), channel.getId(), selectedTime);
                fm.beginTransaction()
                        .replace(R.id.details, fragment)
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .commit();
            } else {
                ((ProgramListFragment) fragment).updatePrograms(selectedTime);
            }
        }
    }

    private void showPopupMenu(View view, int position) {
        Channel channel = recyclerViewAdapter.getItem(position);
        if (activity == null || channel == null) {
            return;
        }

        Program program = appRepository.getProgramData().getItemById(channel.getProgramId());
        Recording recording = appRepository.getRecordingData().getItemByEventId(channel.getProgramId());

        PopupMenu popupMenu = new PopupMenu(activity, view);
        popupMenu.getMenuInflater().inflate(R.menu.program_popup_and_toolbar_menu, popupMenu.getMenu());
        popupMenu.getMenuInflater().inflate(R.menu.external_search_options_menu, popupMenu.getMenu());

        PopupMenuUtil.prepareMenu(activity, popupMenu.getMenu(), program, recording, isNetworkAvailable, htspVersion, isUnlocked);
        PopupMenuUtil.prepareSearchMenu(popupMenu.getMenu(), channel.getProgramTitle(), isNetworkAvailable);
        popupMenu.getMenu().findItem(R.id.menu_play).setVisible(isNetworkAvailable);

        popupMenu.setOnMenuItemClickListener(item -> {
            if (SearchMenuUtils.onMenuSelected(activity, item.getItemId(), channel.getProgramTitle())) {
                return true;
            }
            switch (item.getItemId()) {
                case R.id.menu_record_stop:
                    return menuUtils.handleMenuStopRecordingSelection(recording, null);

                case R.id.menu_record_cancel:
                    return menuUtils.handleMenuCancelRecordingSelection(recording, null);

                case R.id.menu_record_remove:
                    return menuUtils.handleMenuRemoveRecordingSelection(recording, null);

                case R.id.menu_record_once:
                    return menuUtils.handleMenuRecordSelection(channel.getProgramId());

                case R.id.menu_record_once_and_edit:
                    programIdToBeEditedWhenBeingRecorded = channel.getProgramId();
                    return menuUtils.handleMenuRecordSelection(channel.getProgramId());

                case R.id.menu_record_once_custom_profile:
                    return menuUtils.handleMenuCustomRecordSelection(channel.getProgramId(), channel.getId());

                case R.id.menu_record_series:
                    return menuUtils.handleMenuSeriesRecordSelection(channel.getProgramTitle());

                case R.id.menu_play:
                    return menuUtils.handleMenuPlayChannel(channel.getId());

                case R.id.menu_cast:
                    return menuUtils.handleMenuCast("channelId", channel.getId());

                case R.id.menu_add_notification:
                    if (program != null) {
                        return menuUtils.handleMenuAddNotificationSelection(program);
                    } else {
                        return false;
                    }

                default:
                    return false;
            }
        });
        popupMenu.show();
    }

    @Override
    public void onSearchRequested(@NonNull String query) {
        searchQuery = query;
        recyclerViewAdapter.getFilter().filter(query, this);
    }

    @Override
    public boolean onSearchResultsCleared() {
        if (!TextUtils.isEmpty(searchQuery)) {
            Timber.d("Search result not empty, clearing filter and returning true");
            searchQuery = "";
            recyclerViewAdapter.getFilter().filter("", this);
            return true;
        } else {
            Timber.d("Search results empty, returning false");
            return false;
        }
    }

    @NonNull
    @Override
    public String getQueryHint() {
        return getString(R.string.search_channels);
    }

    @Override
    public void onClick(@NonNull View view, int position) {
        //noinspection ConstantConditions
        if ((view.getId() == R.id.icon || view.getId() == R.id.icon_text)
                && Integer.valueOf(sharedPreferences.getString("channel_icon_action", activity.getResources().getString(R.string.pref_default_channel_icon_action))) > 0
                && recyclerViewAdapter.getItemCount() > 0
                && isNetworkAvailable) {
            Channel channel = recyclerViewAdapter.getItem(position);
            menuUtils.handleMenuPlayChannelIcon(channel.getId());

        } else {
            showChannelDetails(position);
        }
    }

    @Override
    public boolean onLongClick(@NonNull View view, int position) {
        showPopupMenu(view, position);
        return true;
    }

    @Override
    public void onFilterComplete(int count) {
        showChannelTagOrChannelCount();
        // Preselect the first result item in the details screen
        if (isDualPane && recyclerViewAdapter.getItemCount() > 0) {
            showChannelDetails(0);
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/channels/ChannelListFragment.kt
package org.tvheadend.tvhclient.ui.features.channels

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.*
import android.widget.Filter
import androidx.appcompat.widget.PopupMenu
import androidx.core.view.children
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import org.tvheadend.data.entity.ChannelTag
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.RecyclerviewFragmentBinding
import org.tvheadend.tvhclient.ui.base.BaseFragment
import org.tvheadend.tvhclient.ui.common.*
import org.tvheadend.tvhclient.ui.common.interfaces.*
import org.tvheadend.tvhclient.ui.features.programs.ProgramListFragment
import org.tvheadend.tvhclient.ui.features.programs.ProgramViewModel
import org.tvheadend.tvhclient.util.extensions.gone
import org.tvheadend.tvhclient.util.extensions.visible
import org.tvheadend.tvhclient.util.extensions.visibleOrGone
import timber.log.Timber

class ChannelListFragment : BaseFragment(), RecyclerViewClickInterface, ChannelTimeSelectedInterface, ChannelTagIdsSelectedInterface, SearchRequestInterface, Filter.FilterListener, ShowProgramListFragmentInterface {

    private lateinit var binding: RecyclerviewFragmentBinding
    private val dialogDismissHandler = Handler(Looper.getMainLooper())
    private var dialogDismissRunnable: Runnable? = null
    private lateinit var programViewModel: ProgramViewModel
    private lateinit var recyclerViewAdapter: ChannelRecyclerViewAdapter
    private lateinit var channelViewModel: ChannelViewModel

    private var channelCount = 0

    // Used in the time selection dialog to show a time entry every x hours.
    private val intervalInHours = 2
    private var programIdToBeEditedWhenBeingRecorded = 0
    private var channelTags: List<ChannelTag> = ArrayList()
    private var selectedTime: Long = 0
    private lateinit var currentTimeUpdateTask: Runnable
    private val currentTimeUpdateHandler = Handler(Looper.getMainLooper())

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = RecyclerviewFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        channelViewModel = ViewModelProvider(requireActivity()).get(ChannelViewModel::class.java)
        programViewModel = ViewModelProvider(requireActivity()).get(ProgramViewModel::class.java)

        arguments?.let {
            channelViewModel.selectedListPosition = it.getInt("listPosition")
            channelViewModel.selectedTimeOffset = it.getInt("timeOffset")
        }

        recyclerViewAdapter = ChannelRecyclerViewAdapter(channelViewModel, isDualPane, this, viewLifecycleOwner)
        binding.recyclerView.layoutManager = LinearLayoutManager(activity)
        binding.recyclerView.adapter = recyclerViewAdapter
        binding.recyclerView.gone()
        binding.recyclerView.setHasFixedSize(true)
        binding.searchProgress.visibleOrGone(baseViewModel.isSearchActive)

        Timber.d("Observing selected time")
        channelViewModel.selectedTime.observe(viewLifecycleOwner,  { time ->
            Timber.d("View model returned selected time $time")
            if (time != null) {
                selectedTime = time
            }
        })

        Timber.d("Observing channel tags")
        channelViewModel.channelTags.observe(viewLifecycleOwner,  { tags ->
            if (tags != null) {
                Timber.d("View model returned ${tags.size} channel tags")
                channelTags = tags
            }
        })

        Timber.d("Observing channels")
        channelViewModel.channels.observe(viewLifecycleOwner,  { channels ->
            if (channels != null) {
                Timber.d("View model returned ${channels.size} channels")
                recyclerViewAdapter.addItems(channels.toMutableList())
                observeSearchQuery()
                observeRecordings()
            }

            binding.recyclerView.visible()
            showStatusInToolbar()
            activity?.invalidateOptionsMenu()

            if (isDualPane && recyclerViewAdapter.itemCount > 0) {
                showProgramListOfSelectedChannelInDualPane(channelViewModel.selectedListPosition)
            }
        })

        channelViewModel.channelCount.observe(viewLifecycleOwner,  { count ->
            channelCount = count
        })

        // Initiate a timer that will update the view model data every minute
        // so that the progress bars will be displayed correctly
        currentTimeUpdateTask = Runnable {
            val currentTime = System.currentTimeMillis()
            Timber.d("Checking if selected time $selectedTime is past current time $currentTime")
            if (!baseViewModel.isSearchActive && selectedTime < currentTime) {
                Timber.d("Updated selected time to current time")
                channelViewModel.setSelectedTime(currentTime)
            }
            currentTimeUpdateHandler.postDelayed(currentTimeUpdateTask, 60000)
        }
    }

    private fun observeRecordings() {
        // Get all recordings for the given channel to check if it belongs to a certain program
        // so the recording status of the particular program can be updated. This is required
        // because the programs are not updated automatically when recordings change.
        Timber.d("Observing recordings")
        channelViewModel.recordings.observe(viewLifecycleOwner,  { recordings ->
            if (recordings != null) {
                Timber.d("View model returned ${recordings.size} recordings")
                recyclerViewAdapter.addRecordings(recordings)
                for (recording in recordings) {
                    // Show the edit recording screen of the scheduled recording
                    // in case the user has selected the record and edit menu item.
                    if (recording.eventId == programIdToBeEditedWhenBeingRecorded && programIdToBeEditedWhenBeingRecorded > 0) {
                        programIdToBeEditedWhenBeingRecorded = 0
                        editSelectedRecording(requireActivity(), recording.id)
                        break
                    }
                }
            }
        })
    }

    private fun observeSearchQuery() {
        Timber.d("Observing search query")
        baseViewModel.searchQueryLiveData.observe(viewLifecycleOwner,  { query ->
            if (query.isNotEmpty()) {
                Timber.d("View model returned search query '$query'")
                onSearchRequested(query)
            } else {
                Timber.d("View model returned empty search query")
                onSearchResultsCleared()
            }
        })
    }

    private fun showStatusInToolbar() {
        // Show either all channels or the name of the selected
        // channel tag and the channel count in the toolbar
        context?.let {
            val toolbarTitle = channelViewModel.getSelectedChannelTagName(it)
            if (!baseViewModel.isSearchActive) {
                toolbarInterface.setTitle(toolbarTitle)
                toolbarInterface.setSubtitle(it.resources.getQuantityString(R.plurals.items, recyclerViewAdapter.itemCount, recyclerViewAdapter.itemCount))
            } else {
                toolbarInterface.setTitle(getString(R.string.search_results))
                toolbarInterface.setSubtitle(it.resources.getQuantityString(R.plurals.channels, recyclerViewAdapter.itemCount, recyclerViewAdapter.itemCount))
            }
        }
    }

    override fun onResume() {
        super.onResume()
        currentTimeUpdateHandler.post(currentTimeUpdateTask)
    }

    override fun onPause() {
        super.onPause()
        currentTimeUpdateHandler.removeCallbacks(currentTimeUpdateTask)
        dialogDismissRunnable?.let { dialogDismissHandler.removeCallbacks(it) }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.channel_list_options_menu, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        val showGenreColors = sharedPreferences.getBoolean("genre_colors_for_channels_enabled", resources.getBoolean(R.bool.pref_default_genre_colors_for_channels_enabled))
        val showChannelTagMenu = sharedPreferences.getBoolean("channel_tag_menu_enabled", resources.getBoolean(R.bool.pref_default_channel_tag_menu_enabled))

        if (!baseViewModel.isSearchActive) {
            menu.findItem(R.id.menu_genre_color_information)?.isVisible = showGenreColors
            menu.findItem(R.id.menu_program_timeframe)?.isVisible = isUnlocked
            menu.findItem(R.id.menu_search)?.isVisible = recyclerViewAdapter.itemCount > 0
            menu.findItem(R.id.menu_search_channels)?.isVisible = recyclerViewAdapter.itemCount > 0

            // Prevent the channel tag menu item from going into the overlay menu
            if (showChannelTagMenu) {
                menu.findItem(R.id.menu_channel_tags)?.setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
            }
        } else {
            menu.findItem(R.id.menu_genre_color_information)?.isVisible = false
            menu.findItem(R.id.menu_program_timeframe)?.isVisible = false
            menu.findItem(R.id.menu_search)?.isVisible = false
            menu.findItem(R.id.menu_search_channels)?.isVisible = false
            menu.findItem(R.id.menu_channel_tags)?.isVisible = false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val ctx = context ?: return super.onOptionsItemSelected(item)
        return when (item.itemId) {
            R.id.menu_channel_tags -> showChannelTagSelectionDialog(ctx, channelTags.toMutableList(), channelCount, this)
            R.id.menu_program_timeframe -> {
                val dialog = showProgramTimeframeSelectionDialog(requireActivity(), channelViewModel.selectedTimeOffset, intervalInHours, 12, this)
                startDialogDismissTimer(dialog)
                true
            }
            R.id.menu_genre_color_information -> showGenreColorDialog(ctx)
            R.id.menu_channel_sort_order -> showChannelSortOrderSelectionDialog(ctx)
            R.id.menu_search_channels -> showSearchForChannelsDialog(ctx)
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun startDialogDismissTimer(dialog: MaterialDialog) {
        dialogDismissRunnable = Runnable {
            if (dialog.isShowing) {
                dialog.dismiss()
            }
        }.also { dialogDismissHandler.postDelayed(it, 60000) }
    }

    private fun showSearchForChannelsDialog(context: Context): Boolean {
        MaterialDialog(context).show {
            title(res = R.string.search_for_channels)
            input(hintRes = R.string.enter_channel_name) { _, text ->
                baseViewModel.startSearchQuery(text.toString())
            }
            positiveButton(R.string.search)
        }
        return true
    }

    override fun onTimeSelected(which: Int) {
        dialogDismissRunnable?.let { dialogDismissHandler.removeCallbacks(it) }
        channelViewModel.selectedTimeOffset = which
        binding.recyclerView.gone()

        // Add the selected list index as extra hours to the current time.
        // If the first index was selected then use the current time.
        var timeInMillis = System.currentTimeMillis()
        timeInMillis += (1000 * 60 * 60 * which * intervalInHours).toLong()
        channelViewModel.setSelectedTime(timeInMillis)
    }

    override fun onChannelTagIdsSelected(ids: Set<Int>) {
        binding.recyclerView.gone()
        channelViewModel.setSelectedChannelTagIds(ids)
    }

    private fun showChannelDetails(position: Int) {
        if (!isDualPane) {
            showProgramListOfSelectedChannelInSinglePane(position)
        } else {
            showProgramListOfSelectedChannelInDualPane(position)
        }
    }

    private fun showProgramListOfSelectedChannelInSinglePane(position: Int) {
        channelViewModel.selectedListPosition = position
        recyclerViewAdapter.setPosition(position)
        val channel = recyclerViewAdapter.getItem(position)
        if (channel == null || !isVisible) {
            return
        }

        // Show the fragment to display the program list of the selected channel.
        val fragment = ProgramListFragment.newInstance(channel.name ?: "", channel.id, selectedTime)
        activity?.supportFragmentManager?.beginTransaction()?.also {
            it.replace(R.id.main, fragment)
            it.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
            it.addToBackStack(null)
            it.commit()
        }
    }

    private fun showProgramListOfSelectedChannelInDualPane(position: Int) {
        channelViewModel.selectedListPosition = position
        recyclerViewAdapter.setPosition(position)
        val channel = recyclerViewAdapter.getItem(position)
        if (channel == null || !isVisible) {
            return
        }

        Timber.d("Showing program list for selected channel")
        // Check if an instance of the program list fragment for the selected channel is
        // already available. If an instance exist already then update the selected time
        // that was selected from the channel list.
        val fm = activity?.supportFragmentManager
        var fragment = fm?.findFragmentById(R.id.details)
        if (fragment !is ProgramListFragment
                || programViewModel.channelId != channel.id) {
            Timber.d("Channel has changed, showing new program list")
            fragment = ProgramListFragment.newInstance(channel.name ?: "", channel.id, selectedTime)

            // Check the lifecycle state to avoid committing the transaction
            // after the onSaveInstance method was already called which would
            // trigger an illegal state exception.
            if (lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) {
                fm?.beginTransaction()?.also {
                    it.replace(R.id.details, fragment)
                    it.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                    it.commit()
                }
            }
        } else {
            Timber.d("Channel is the same, updating only time in the program list")
            programViewModel.selectedTime = selectedTime
        }
    }

    private fun showPopupMenu(view: View, position: Int) {
        val channel = recyclerViewAdapter.getItem(position) ?: return
        val ctx = context ?: return

        val program = channelViewModel.getProgramById(channel.programId)
        val recording = channelViewModel.getRecordingById(channel.programId)

        val popupMenu = PopupMenu(ctx, view)
        popupMenu.menuInflater.inflate(R.menu.program_popup_and_toolbar_menu, popupMenu.menu)
        popupMenu.menuInflater.inflate(R.menu.external_search_options_menu, popupMenu.menu)

        preparePopupOrToolbarRecordingMenu(ctx, popupMenu.menu, recording, isConnectionToServerAvailable, htspVersion, isUnlocked)
        preparePopupOrToolbarSearchMenu(popupMenu.menu, channel.programTitle, isConnectionToServerAvailable)
        preparePopupOrToolbarMiscMenu(ctx, popupMenu.menu, program, isConnectionToServerAvailable, isUnlocked)

        // If no program data is available for the channel, hide all menu items except
        // playing the channel. This is the only option possible
        if (program == null && isConnectionToServerAvailable) {
            popupMenu.menu.children.forEach { it.isVisible = false }
            popupMenu.menu.findItem(R.id.menu_play)?.isVisible = true
        }

        popupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.menu_stop_recording -> return@setOnMenuItemClickListener showConfirmationToStopSelectedRecording(ctx, recording, null)
                R.id.menu_cancel_recording -> return@setOnMenuItemClickListener showConfirmationToCancelSelectedRecording(ctx, recording, null)
                R.id.menu_remove_recording -> return@setOnMenuItemClickListener showConfirmationToRemoveSelectedRecording(ctx, recording, null)
                R.id.menu_record_program -> return@setOnMenuItemClickListener recordSelectedProgram(ctx, channel.programId, channelViewModel.getRecordingProfile(), htspVersion)
                R.id.menu_record_program_and_edit -> {
                    programIdToBeEditedWhenBeingRecorded = channel.programId
                    return@setOnMenuItemClickListener recordSelectedProgram(ctx, channel.programId, channelViewModel.getRecordingProfile(), htspVersion)
                }
                R.id.menu_record_program_with_custom_profile -> return@setOnMenuItemClickListener recordSelectedProgramWithCustomProfile(ctx, channel.programId, channel.id, channelViewModel.getRecordingProfileNames(), channelViewModel.getRecordingProfile())
                R.id.menu_record_program_as_series_recording -> return@setOnMenuItemClickListener recordSelectedProgramAsSeriesRecording(ctx, channel.programTitle, channel.id, channelViewModel.getRecordingProfile(), htspVersion)
                R.id.menu_play -> return@setOnMenuItemClickListener playSelectedChannel(ctx, channel.id, isUnlocked)
                R.id.menu_cast -> return@setOnMenuItemClickListener castSelectedChannel(ctx, channel.id)

                R.id.menu_search_imdb -> return@setOnMenuItemClickListener searchTitleOnImdbWebsite(ctx, channel.programTitle)
                R.id.menu_search_fileaffinity -> return@setOnMenuItemClickListener searchTitleOnFileAffinityWebsite(ctx, channel.programTitle)
                R.id.menu_search_youtube -> return@setOnMenuItemClickListener searchTitleOnYoutube(ctx, channel.programTitle)
                R.id.menu_search_google -> return@setOnMenuItemClickListener searchTitleOnGoogle(ctx, channel.programTitle)
                R.id.menu_search_epg -> return@setOnMenuItemClickListener searchTitleInTheLocalDatabase(requireActivity(), baseViewModel, channel.programTitle, channel.id)

                R.id.menu_add_notification -> return@setOnMenuItemClickListener addNotificationProgramIsAboutToStart(ctx, program, channelViewModel.getRecordingProfile())
                else -> return@setOnMenuItemClickListener false
            }
        }
        popupMenu.show()
    }

    override fun onSearchRequested(query: String) {
        recyclerViewAdapter.filter.filter(query, this)
    }

    override fun onSearchResultsCleared() {
        recyclerViewAdapter.filter.filter("", this)
    }

    override fun getQueryHint(): String {
        return getString(R.string.search_programs)
    }

    override fun onClick(view: View, position: Int) {
        if ((view.id == R.id.icon || view.id == R.id.icon_text)
                && Integer.valueOf(sharedPreferences.getString("channel_icon_action", resources.getString(R.string.pref_default_channel_icon_action))!!) > 0
                && isConnectionToServerAvailable) {
            recyclerViewAdapter.getItem(position)?.let {
                playOrCastChannel(view.context, it.id, isUnlocked)
            }
        } else {
            showChannelDetails(position)
        }
    }

    override fun onLongClick(view: View, position: Int): Boolean {
        showPopupMenu(view, position)
        return true
    }

    override fun onFilterComplete(count: Int) {
        binding.searchProgress.gone()
        showStatusInToolbar()
        // Show the first search result item in the details screen
        if (isDualPane) {
            when {
                recyclerViewAdapter.itemCount > channelViewModel.selectedListPosition -> {
                    showChannelDetails(channelViewModel.selectedListPosition)
                }
                recyclerViewAdapter.itemCount <= channelViewModel.selectedListPosition -> {
                    showChannelDetails(0)
                }
                recyclerViewAdapter.itemCount == 0 -> {
                    removeDetailsFragment()
                }
            }
        }
    }
}
